curp/src/keybased_channel.rs

```diff
@@ -0,0 +1,297 @@
+    /// Insert a key and message into the pending pool
+    fn insert_pending(&mut self, key: K, msg: M) {
+        let entry = self.pending.entry(key).or_insert_with(|| VecDeque::new());
+        entry.push_back(msg);
+    }
+
+    /// Insert a key and message to the inner queue
+    fn insert_inner(&mut self, key: K, msg: M) {
+        self.inner.push_back((key, msg));
+    }
+
+    /// Append a key and message to this `KeybasedChannel`
+    fn append(&mut self, key: K, msg: M) {
+        if self.keys.contains(&key) {
+            self.insert_pending(key, msg);
+        } else {
+            let _old = self.keys.insert(key.clone());
+            self.insert_inner(key, msg);
+            self.new_msg_event.notify(1);
+        }
+    }
+
+    /// Move a message for the `key` from pending to inner
+    fn move_pending_to_inner(&mut self, key: &K) {
+        let msg = self
+            .pending
+            .get_mut(key)
+            .map(|msgs| (msgs.pop_front(), msgs.is_empty()));
+
+        if let Some((option_m, need_clean)) = msg {
+            if let Some(m) = option_m {
+                self.insert_inner(key.clone(), m);
+                self.new_msg_event.notify(1);
+            }
+
+            if need_clean {
+                // remove the key from the pending if the key's queue is empty
+                let _empty_queue = self.pending.remove(key);
+            }
+        }
+    }
+
+    /// Try to receive a message from this Queue.
+    /// Return `None` if there's nothing in the `inner`, otherwise
+    /// return `Some(message)`
+    fn try_recv(&mut self) -> Option<M> {
+        self.inner.pop_front().map(|(key, msg)| {
+            if !self.pending.contains_key(&key) {
+                let _key = self.keys.remove(&key);
+            }
+            msg
+        })
+    }
+}
+
+/// The Sender for the `KeybasedChannel`
+pub(crate) struct KeybasedChannelSender<K: Eq + Hash, M> {
+    /// The channel
+    channel: Arc<Mutex<KeybasedChannel<K, M>>>,
+}
+
+impl<K: Eq + Hash + Clone, M> KeybasedChannelSender<K, M> {
+    /// Send a key and message to the channel
+    #[allow(dead_code)]
+    pub(crate) fn send(&self, key: K, msg: M) {
+        let mut channel = self.channel.lock();
+        channel.append(key, msg);
+    }
+}
+
+/// The Receiver for the `KeybasedChannel`
+pub(crate) struct KeybasedChannelReceiver<K: Eq + Hash, M> {
+    /// The channel
+    channel: Arc<Mutex<KeybasedChannel<K, M>>>,
+    /// The join handler for the task, which handles the message complete event.
+    handle: JoinHandle<()>,
+}
+
+impl<K: Eq + Hash, M> Drop for KeybasedChannelReceiver<K, M> {
+    fn drop(&mut self) {
+        self.handle.abort();
+    }
+}
+
+impl<K: Eq + Hash + Clone, M> KeybasedChannelReceiver<K, M> {
+    /// Receive a message
+    /// Return (message, `msg_complete_sender`)
+    #[allow(dead_code)]
+    pub(crate) fn recv(&self) -> (M, UnboundedSender<K>) {
+        loop {
+            let lisenter = {
```

Comment:
possible to move the common logic to a helper function?

Reply:
I've taken it into consideration, but there's logic to return the value in the middle of a loop, line 123. A function can't work that way.